runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/processing/StreamingCommitFinalizer.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.work.processing;
+
+import java.time.Duration;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.beam.runners.dataflow.worker.util.BoundedQueueExecutor;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.cache.Cache;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.cache.CacheBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class StreamingCommitFinalizer {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingCommitFinalizer.class);
+  private static final Duration DEFAULT_CACHE_ENTRY_EXPIRY = Duration.ofMinutes(5L);
+  private final Cache<Long, Runnable> onCommitFinalizedCache;
+  private final BoundedQueueExecutor workExecutor;
+
+  private StreamingCommitFinalizer(
+      Cache<Long, Runnable> onCommitFinalizedCache, BoundedQueueExecutor workExecutor) {
+    this.onCommitFinalizedCache = onCommitFinalizedCache;
+    this.workExecutor = workExecutor;
+  }
+
+  static StreamingCommitFinalizer create(BoundedQueueExecutor workExecutor) {
+    return new StreamingCommitFinalizer(
+        CacheBuilder.newBuilder().expireAfterWrite(DEFAULT_CACHE_ENTRY_EXPIRY).build(),
+        workExecutor);
+  }
+
+  /**
+   * Stores a map of user worker generated id's and callbacks to execute once a commit has been
```

Comment:
how about being more specific "user worker generated finalization ids"
since we have lots of ids floating around :)